src/distribute_kv_cache/rpc/workerpool.rs

```diff
@@ -0,0 +1,223 @@
+    max_waiting_jobs: usize,
+    /// The job queue for the worker pool, with a maximum buffer capacity of `max_waiting_jobs`.
+    /// 1. When the job queue is full, the worker pool will block to receive new jobs.
+    /// 2. When sender is dropped, the receiver will return `None` and the worker pool will shutdown.
+    /// 3. Check the receiver reference count to determine if the worker pool is still alive.
+    job_sender: Arc<flume::Sender<JobImpl>>,
+    /// The workers in the worker pool.
+    worker_queue: Vec<Worker>,
+}
+
+impl Debug for WorkerPool {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("WorkerPool")
+            .field("max_workers", &self.max_workers)
+            .field("max_waiting_jobs", &self.max_waiting_jobs)
+            .field("worker_queue_len", &self.worker_queue.len())
+            .finish_non_exhaustive()
+    }
+}
+
+impl WorkerPool {
+    /// Create a new worker pool, which contains a number of max workers and max waiting jobs.
+    #[must_use]
+    pub fn new(max_workers: usize, max_waiting_jobs: usize) -> Self {
+        let (job_sender, job_receiver) = flume::bounded::<JobImpl>(max_waiting_jobs);
+        let mut worker_queue = Vec::new();
+
+        // In current implementation, we create a fixed number of workers.
+        let receiver = Arc::new(job_receiver);
+        for _ in 0..max_workers {
+            let worker = Worker::new(Arc::clone(&receiver));
+            worker.start();
+            worker_queue.push(worker);
+        }
+
+        Self {
+            max_workers,
+            max_waiting_jobs,
+            job_sender: Arc::new(job_sender),
+            worker_queue,
+        }
+    }
+
+    /// Submit a job to the worker pool synchronously, and block until the job is completed.
+    /// Other process will be blocked until the job is completed.
+    /// If all job try to submit the job, the process will be blocked.
+    pub fn submit_job(&self, job: JobImpl) -> Result<(), RpcError> {
+        // Submit the job to the job queue.
+        self.job_sender
+            .send(job)
+            .map_err(|_foo| RpcError::InternalError("Failed to submit job".to_owned()))?;
+
+        Ok(())
+    }
+
+    /// Shutdown
+    ///
+    /// Drop the worker, and the worker will exit.
+    pub fn shutdown(&self) {
+        for worker in &self.worker_queue {
+            worker.exit();
+        }
+    }
+}
+
+/// A worker that can execute async jobs.
+#[allow(dead_code)]
+#[derive(Clone)]
+struct Worker {
+    /// Job recv channel
+    job_rx: Arc<flume::Receiver<JobImpl>>,
+    /// Shutdown recv channel
+    shutdown_rx: flume::Receiver<()>,
+    /// Shutdown send channel
+    shutdown_tx: flume::Sender<()>,
+}
+
+impl Worker {
+    /// Create a new worker.
+    fn new(receiver: Arc<flume::Receiver<JobImpl>>) -> Self {
+        debug!("Create a new worker...");
+        let (shutdown_tx, shutdown_rx) = flume::bounded::<()>(1);
```

Comment:
It's a event notify. Maybe event-listener is a better choice.

Reply:
Changed to `tokio::sync::Notify` for better understand.